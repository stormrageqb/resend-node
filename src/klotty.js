import pkg from '../package.json'

export class Klotty {
  constructor(apiKey) {
    if (!apiKey) {
      throw 'Missing API key. Pass it to the constructor `new Klotty("kl_123")`'
    }

    this.apiKey = apiKey
    this.baseURL = 'https://api.klotty.com'
    this.headers = {
      'Authorization': `Bearer ${this.apiKey}`,
      'User-Agent': `node/${pkg.version}`,
      'Content-Type': 'application/json',
    }
  }

  async sendEmail(body = {}) {
    const req = await fetch(`${this.baseURL}/email`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({
        from: body.from,
        to: body.to,
        subject: body.subject,
        html: body.html,
        text: body.text,
      })
    })

    return await req.json()
  }
}