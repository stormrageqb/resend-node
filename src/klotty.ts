import axios, { AxiosInstance } from 'axios';
import { render } from '@react-email/render';
import { SendEmailData } from './interfaces';
import { version } from '../package.json';

export class Klotty {
  readonly baseUrl: string;
  private readonly headers: HeadersInit;
  private readonly request: AxiosInstance;

  constructor(readonly apiKey?: string) {
    if (!apiKey) {
      this.apiKey = process.env.KLOTTY_API_KEY;

      if (!this.apiKey) {
        throw new Error(
          'Missing API key. Pass it to the constructor `new Klotty("kl_123")`',
        );
      }
    }

    this.apiKey = apiKey;
    this.baseUrl = process.env.KLOTTY_BASE_URL || 'https://api.klotty.com';
    this.headers = {
      Authorization: `Bearer ${this.apiKey}`,
      'User-Agent': `node:${version}`,
      'Content-Type': 'application/json',
    };
    this.request = axios.create({
      baseURL: this.baseUrl,
      headers: this.headers,
    });
  }

  async sendEmail(data: SendEmailData): Promise<object> {
    try {
      const path = `${this.baseUrl}/email`;

      if (data.react) {
        data.html = render(data.react);
        delete data.react;
      }

      const response = await this.request(path, { method: 'POST', data });
      return response.data;
    } catch (error) {
      throw error;
    }
  }
}
