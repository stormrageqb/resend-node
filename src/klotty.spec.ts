import { Klotty } from './klotty';
import MockAdapater from 'axios-mock-adapter';
import axios from 'axios';

const mock = new MockAdapater(axios);
const klotty = new Klotty('kl_924b3rjh2387fbewf823');

describe('Klotty', () => {
  afterEach(() => mock.resetHistory());

  it('throws API key error', () => {
    expect(() => new Klotty()).toThrowErrorMatchingSnapshot();
  });

  it('sends email', async () => {
    const payload = {
      from: 'bu@klotty.com',
      to: 'zeno@klotty.com',
      subject: 'Hello World',
    };
    mock.onPost('https://api.klotty.com/email', payload).replyOnce(200, {
      id: '1234',
      from: 'bu@klotty.com',
      to: 'zeno@klotty.com',
      created_at: '123',
    });

    const data = await klotty.sendEmail(payload);
    expect(data).toMatchSnapshot();
  });

  it('sends email with multiple recipients', async () => {
    const payload = {
      from: 'admin@klotty.com',
      to: ['bu@klotty.com', 'zeno@klotty.com'],
      subject: 'Hello World',
    };
    mock.onPost('https://api.klotty.com/email', payload).replyOnce(200, {
      id: '1234',
      from: 'admin@klotty.com',
      to: ['bu@klotty.com', 'zeno@klotty.com'],
      created_at: '123',
    });

    const data = await klotty.sendEmail(payload);
    expect(data).toMatchSnapshot();
  });

  it('sends email with multiple bcc recipients', async () => {
    const payload = {
      from: 'admin@klotty.com',
      to: 'bu@klotty.com',
      bcc: ['foo@klotty.com', 'bar@klotty.com'],
      subject: 'Hello World',
    };
    mock.onPost('https://api.klotty.com/email', payload).replyOnce(200, {
      id: '1234',
      from: 'admin@klotty.com',
      to: 'bu@klotty.com',
      bcc: ['foo@klotty.com', 'bar@klotty.com'],
      created_at: '123',
    });

    const data = await klotty.sendEmail(payload);
    expect(data).toMatchSnapshot();
  });

  it('sends email with multiple cc recipients', async () => {
    const payload = {
      from: 'admin@klotty.com',
      to: 'bu@klotty.com',
      cc: ['foo@klotty.com', 'bar@klotty.com'],
      subject: 'Hello World',
    };
    mock.onPost('https://api.klotty.com/email', payload).replyOnce(200, {
      id: '1234',
      from: 'admin@klotty.com',
      to: 'bu@klotty.com',
      cc: ['foo@klotty.com', 'bar@klotty.com'],
      created_at: '123',
    });

    const data = await klotty.sendEmail(payload);
    expect(data).toMatchSnapshot();
  });
});
