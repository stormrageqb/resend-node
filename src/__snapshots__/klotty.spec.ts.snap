// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Klotty sends email 1`] = `
Object {
  "created_at": "123",
  "from": "bu@klotty.com",
  "id": "1234",
  "to": "zeno@klotty.com",
}
`;

exports[`Klotty sends email with multiple bcc recipients 1`] = `
Object {
  "bcc": Array [
    "foo@klotty.com",
    "bar@klotty.com",
  ],
  "created_at": "123",
  "from": "admin@klotty.com",
  "id": "1234",
  "to": "bu@klotty.com",
}
`;

exports[`Klotty sends email with multiple cc recipients 1`] = `
Object {
  "cc": Array [
    "foo@klotty.com",
    "bar@klotty.com",
  ],
  "created_at": "123",
  "from": "admin@klotty.com",
  "id": "1234",
  "to": "bu@klotty.com",
}
`;

exports[`Klotty sends email with multiple recipients 1`] = `
Object {
  "created_at": "123",
  "from": "admin@klotty.com",
  "id": "1234",
  "to": Array [
    "bu@klotty.com",
    "zeno@klotty.com",
  ],
}
`;

exports[`Klotty throws API key error 1`] = `"Missing API key. Pass it to the constructor \`new Klotty(\\"kl_123\\")\`"`;
